File: 0906/보석 쇼핑.swift
// 투 포인터 알고리즘은 파악할 수 있었지만 start, end를 잡는 것 밑 검사를 하는 과정에서의 서순에 따라서 각 루프별
// start, end의 값의 상태가 바뀔 수 있었고, end == gems.count일 때의 엣지 케이스에서
// start를 1씩 늘려 가며 검사하지 않고 그대로 종료되는 엣지케이스를 생각하지 못했었다.

import Foundation

func solution(_ gems:[String]) -> [Int] {
    var gemDict: [String:Bool] = [:]
    var gemCount = 0
    for gem in gems {
        if gemDict[gem] == nil {
            gemCount += 1
            gemDict[gem] = true
        }
    }
    
    var start = 0
    var end = 0
    var currentDict: [String:Int] = [:]
    var currentCount = 0
    
    var answer = [0, gems.count]
    
    while start <= end && end < gems.count {
        if currentCount < gemCount {
            let gem = gems[end]
            
            if let _ = currentDict[gem] {
                currentDict[gem]! += 1
            } else {
                currentDict[gem] = 1
                currentCount += 1
            }
            
            end += 1
        } else if currentCount == gemCount {
            let gem = gems[start]
            
            if currentDict[gem]! > 1 {
                currentDict[gem]! -= 1
            } else if currentDict[gem] == 1{
                currentDict[gem] = nil
                currentCount -= 1
            }
            
            start += 1
        }
        
        if currentCount == gemCount {
            if (answer[1]-answer[0] == end-start-1 && answer[0] > start+1) || (answer[1]-answer[0] > end-start-1) {
                answer = [start+1, end]
            }
        }
    }
    
    while end == gems.count && currentCount == gemCount {
        let gem = gems[start]
        
        if currentDict[gem]! > 1 {
            currentDict[gem]! -= 1
        } else if currentDict[gem] == 1{
            currentDict[gem] = nil
            currentCount -= 1
        }
        
        start += 1
        
        if currentCount == gemCount {
            if (answer[1]-answer[0] == end-start-1 && answer[0] > start+1) || (answer[1]-answer[0] > end-start-1) {
                answer = [start+1, end]
            }
        }

    }
    
    return answer
}

File: 0906/보석 쇼핑.py
def solution(gems):
    answer = []
    
    # 보석의 종류
    n = len(set(gems))    
    
    # 각 보석이 등장한 최근위치
    order = {}
    
    # 보석의 첫 종류
    first = []
    
    # 구매할 구간의 시작점
    start = 0
    
    # 현재까지 구매한 종류를 체크하기 위한 set
    cart = set()
    
    for idx, gem in enumerate(gems):
        
        # 등장한 위치를 기록
        order[gem] = idx
        # 구매 목록에 추가
        cart.add(gem)
        
        # 구매 목록이 종류와 같아지면 우선 기록
        # 답이 아닌 이유? 먼저 완성된다고 해도 뒤에 더 짧은 게 나오면 그게 답이기 때문
        # 우선순위가 길이 -> 등장 순인데 등장 -> 길이 순으로 풀어서 처음에 틀렸음
        if len(cart) == n:
            answer.append([start+1, idx+1])
            
        # 첫 보석을 기록
        if not first:
            first.append(gem)
        
        # 첫 보석과 같은 보석이 등장할 때
        if first:
            if first[0] == gem:
                next_gem = min(order, key=order.get)
                first = [next_gem]
                start = order[next_gem]
                
                # 카트 목록 갱신
                cart = set(gems[start:idx+1])
                if len(cart) == n:
                    answer.append([start+1, idx+1])
        
    # 짧은 것 중 가장 먼저 등장한 구간
    answer = sorted(answer, key=lambda x:[x[1]-x[0], x[0]])
    return answer[0]